### ISSUE TITLE
Intermittent results for pvget -r 

### ISSUE THREAD

**Issue by bhill-slac** [mdavidsaver/p4p#61]
We have a group using python to sample PVA images.
They first request the dimension field so they can configure their code to handle the image,
but often these requests fail to provide the dimension field.

When it fails, it looks like this:
% pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image
SL1K2:EXIT:CAM:DATA1:Pva:Image  %

When it succeeds, it always returns all the fields, not just the requested dimension.
% pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image
SL1K2:EXIT:CAM:DATA1:Pva:Image epics:nt/NTNDArray:1.0 
    union value
        (none)
    codec_t codec
        string name 
        any parameters
            (none)
    long compressedSize 0
    long uncompressedSize 0
    dimension_t[] dimension
        dimension_t 
            int size 2048
            int offset 0
            int fullSize 2048
            int binning 1
            boolean reverse false
        dimension_t 
            int size 2048
            int offset 0
            int fullSize 2048
            int binning 1
            boolean reverse false
    int uniqueId 0
    time_t dataTimeStamp <undefined>              
        long secondsPastEpoch 0
        int nanoseconds 0
        int userTag 0
    epics:nt/NTAttribute:1.0[] attribute
    string descriptor 
    alarm_t alarm 
        int severity 0
        int status 0
        string message 
    time_t timeStamp <undefined>              
        long secondsPastEpoch 0
        int nanoseconds 0
        int userTag 0
    display_t display
        double limitLow 0
        double limitHigh 0
        string description 
        string format 
        string units 
%

Here's an example of a pvget request for the dimension of a Pva:Image on the rix subnet run multiple 
times from drp-neh-cmp015.
This works every time on rix-console as it gets the info directly from the IOC instead of through the PVA gateway.

% for ((i=0;i<20;i++)); do echo pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image \| egrep size; pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image  | egrep -w size; sleep 10; done
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
            int size 2048
            int size 2048
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
            int size 2048
            int size 2048
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
            int size 2048
            int size 2048
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
pvget -r dimension SL1K2:EXIT:CAM:DATA1:Pva:Image | egrep size
            int size 2048
            int size 2048
%

**Comment by mdavidsaver** [comment]
> They first request the dimension field so they can configure their code to handle the image,

This seems indicative of some underlying opportunities for improvement.

**Comment by mdavidsaver** [comment]
At present the gateway is ignoring most of the downstream pvRequest when creating the upstream pvRequest.  So the gateway itself will always fetch the whole value.

https://github.com/mdavidsaver/p4p/blob/de508a13edfd440470b5f2b97196278f61131c11/src/gwchannel.cpp#L680-L689

However, the downstream pvRequest should still be applying to the result going back to the downstream client.

https://github.com/mdavidsaver/p4p/blob/de508a13edfd440470b5f2b97196278f61131c11/src/gwchannel.cpp#L554-L559

I'm not sure why this would be intermittent though.

**Comment by mdavidsaver** [comment]
fyi. A client with the `UNCACHED`ACF permission can bypass most of the gateway logic.  Can you test:

```
pvget -r 'field(dimension)record[cache=false]' ...
```

**Comment by bhill-slac** [comment]
% pvget -r 'field(dimension)record[cache=false]' SL1K2:EXIT:CAM:DATA1:Pva:Image 
SL1K2:EXIT:CAM:DATA1:Pva:Image Error Gateway disallows uncachable get

**Comment by mdavidsaver** [comment]
Right, this client is missing the `UNCACHED` permission, which isn't granted by default.  You would need to add a `RULE(1, UNCACHED)`.

**Comment by mdavidsaver** [comment]
@bhill-slac Was adding `UNCACHED` enough?  Or where you able to prevail upon this user to switch to using a single subscription?

**Comment by echandler-anl** [comment]
I also am looking at using pvget -r with the PVA gateway.  I tried using the following ACF file:

```
ASG(DEFAULT) {
 RULE(1, UNCACHED)
}
```

When I run `pvget -r 'field(FIELDNAME)record[cache=false]'` it still gets the "Gateway disallows uncachable get" error.

**Comment by mdavidsaver** [comment]
@echandler-anl Does your gateway configuration enable the internal [status PVs](https://epics-base.github.io/p4p/gw.html#status-pvs)?

The `<statusprefix>asTest` RPC call is particularly relevant when testing ACF configuration.  eg. by default the UNCACHED permission is not granted.

```
$ pvxcall sts:asTest pv=cnt user=bar
...
       bool uncached = false
...
```

With no PV list file, and loading only:

```
ASG(DEFAULT)
{
        RULE(1,WRITE,TRAPWRITE)
        RULE(1,UNCACHED)
}
```

Should grant all permission, and enable audit logging, for all PVs.

```
$ pvxcall sts:asTest pv=cnt user=bar
...
        bool put = true
        bool rpc = true
        bool uncached = true
        bool audit = true

...
```

Then an uncached GET is permitted.

```
$ pvxget -r 'record[cache=false]' cnt
cnt
    value uint32_t = 22
...
```
